src/modules/postcode-lookup/index.js

```diff
@@ -0,0 +1,171 @@
+
+    function addSearchResultsToSelectElement(data) {
+        const selectElementResults = window.document.getElementById(
+            'address-search-results-dropdown'
+        );
+        // Clear the results drop-down list.
+        const options = window.document.querySelectorAll('#address-search-results-dropdown option');
+        options.forEach(o => o.remove());
+
+        // Update the JSON object with the data results.
+        tmpAddressSearchResultsJson = data.results;
+        const option = window.document.createElement('option');
+        option.text =
+            data.header.totalresults === 1
+                ? '1 address found'
+                : `${data.header.totalresults} addresses found`;
+        selectElementResults.add(option);
+
+        // See also {@link https://apidocs.os.uk/reference/os-places-dpa-output}
+        const dataset = 'DPA';
+        // Loop through the data results; adding the address string as a new option
+        // to the results drop-down list.
+        data.results.forEach(function addElemets(val, i) {
+            const opt = window.document.createElement('option');
+            opt.value = i;
+            opt.text = val[dataset.toUpperCase()].ADDRESS;
+            selectElementResults.add(opt);
+        });
+
+        // Display the hidden search results div
+        const searchResultsDiv = window.document.getElementById('address-search-results');
+        searchResultsDiv.style.display = 'block';
+    }
+
+    async function addressSearch() {
+        const postcodeInput = window.document.getElementById('postcode-search-input').value;
+
+        const response = await fetch(`/address-finder/postcode?postcode=${postcodeInput}`);
+        if (!response.ok) {
+            // throw new Error(`HTTP error! status: ${response.status}`);
+            // TODO proper error handling
+            const msg = 'TBC API Error Status error handling.';
+            throw msg;
+        }
+
+        const data = await response.json();
+
+        // TODO add proper error handling for no results found
+        if (data.header.totalresults === 0 || !data.results) {
+            const msg = 'TBC No matching results found.';
+            throw msg;
+        }
+
+        addSearchResultsToSelectElement(data);
+    }
+
+    function createContentElements() {
+        // CREATE CONTENT ELEMENT
+        const descriptionDiv = window.document.createElement('div');
+        descriptionDiv.appendChild(
+            window.document.createTextNode(
+                'Enter a UK postcode to search for your address, or fill out the fields manually.'
+            )
+        );
+        descriptionDiv.setAttribute('id', 'fill-out-the-fields-manually-hint');
+        descriptionDiv.setAttribute('class', 'govuk-hint');
+        window.document
+            .querySelector('.govuk-form-group')
+            .insertAdjacentElement('beforeBegin', descriptionDiv);
+    }
+
+    function createPostcodeSearchElements() {
```

Comment:
Would it be easier to just have a chunk of HTML that you append? i.e:

const postcodeDiv = window.document.createElement('div');
postcodeDiv.innerHTML = '<div id="postcode-search" class="govuk-form-group">..All other html in string literal form..</div>';
window.document
            .getElementById('fill-out-the-fields-manually-hint').insertAdjacentElement(postcodeDiv);

Reply:
See the Security considerations [here](https://developer.mozilla.org/en-US/docs/Web/API/Element/innerHTML)